// @ark-ui/react/range-slider/index.mjs
'use client';
import { RangeSlider as RangeSlider$1 } from './range-slider.mjs';
export { useRangeSliderContext } from './range-slider-context.mjs';
import { RangeSliderControl } from './range-slider-control.mjs';
import { RangeSliderLabel } from './range-slider-label.mjs';
import { RangeSliderMarker } from './range-slider-marker.mjs';
import { RangeSliderMarkerGroup } from './range-slider-marker-group.mjs';
import { RangeSliderOutput } from './range-slider-output.mjs';
import { RangeSliderRange } from './range-slider-range.mjs';
import { RangeSliderThumb } from './range-slider-thumb.mjs';
import { RangeSliderTrack } from './range-slider-track.mjs';
export { anatomy as rangeSliderAnatomy } from '@zag-js/range-slider';

const RangeSlider = Object.assign(RangeSlider$1, {
  Root: RangeSlider$1,
  Control: RangeSliderControl,
  Label: RangeSliderLabel,
  Marker: RangeSliderMarker,
  MarkerGroup: RangeSliderMarkerGroup,
  Output: RangeSliderOutput,
  Range: RangeSliderRange,
  Thumb: RangeSliderThumb,
  Track: RangeSliderTrack
});

export { RangeSlider, RangeSliderControl, RangeSliderLabel, RangeSliderMarker, RangeSliderMarkerGroup, RangeSliderOutput, RangeSliderRange, RangeSliderThumb, RangeSliderTrack };
