// @ark-ui/react/range-slider/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const rangeSlider$1 = require('./range-slider.cjs');
const rangeSliderContext = require('./range-slider-context.cjs');
const rangeSliderControl = require('./range-slider-control.cjs');
const rangeSliderLabel = require('./range-slider-label.cjs');
const rangeSliderMarker = require('./range-slider-marker.cjs');
const rangeSliderMarkerGroup = require('./range-slider-marker-group.cjs');
const rangeSliderOutput = require('./range-slider-output.cjs');
const rangeSliderRange = require('./range-slider-range.cjs');
const rangeSliderThumb = require('./range-slider-thumb.cjs');
const rangeSliderTrack = require('./range-slider-track.cjs');
const rangeSlider = require('@zag-js/range-slider');

const RangeSlider = Object.assign(rangeSlider$1.RangeSlider, {
  Root: rangeSlider$1.RangeSlider,
  Control: rangeSliderControl.RangeSliderControl,
  Label: rangeSliderLabel.RangeSliderLabel,
  Marker: rangeSliderMarker.RangeSliderMarker,
  MarkerGroup: rangeSliderMarkerGroup.RangeSliderMarkerGroup,
  Output: rangeSliderOutput.RangeSliderOutput,
  Range: rangeSliderRange.RangeSliderRange,
  Thumb: rangeSliderThumb.RangeSliderThumb,
  Track: rangeSliderTrack.RangeSliderTrack
});

exports.useRangeSliderContext = rangeSliderContext.useRangeSliderContext;
exports.RangeSliderControl = rangeSliderControl.RangeSliderControl;
exports.RangeSliderLabel = rangeSliderLabel.RangeSliderLabel;
exports.RangeSliderMarker = rangeSliderMarker.RangeSliderMarker;
exports.RangeSliderMarkerGroup = rangeSliderMarkerGroup.RangeSliderMarkerGroup;
exports.RangeSliderOutput = rangeSliderOutput.RangeSliderOutput;
exports.RangeSliderRange = rangeSliderRange.RangeSliderRange;
exports.RangeSliderThumb = rangeSliderThumb.RangeSliderThumb;
exports.RangeSliderTrack = rangeSliderTrack.RangeSliderTrack;
Object.defineProperty(exports, 'rangeSliderAnatomy', {
  enumerable: true,
  get: () => rangeSlider.anatomy
});
exports.RangeSlider = RangeSlider;
