// @ark-ui/react/radio-group/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const radio = require('./radio.cjs');
const radioContext = require('./radio-context.cjs');
const radioControl = require('./radio-control.cjs');
const radioGroup = require('./radio-group.cjs');
const radioGroupContext = require('./radio-group-context.cjs');
const radioGroupLabel = require('./radio-group-label.cjs');
const radioLabel = require('./radio-label.cjs');
const radio$1 = require('@zag-js/radio-group');

const RadioGroup = Object.assign(radioGroup.RadioGroup, {
  Root: radioGroup.RadioGroup,
  Label: radioGroupLabel.RadioGroupLabel,
  Radio: radio.Radio,
  RadioLabel: radioLabel.RadioLabel,
  RadioControl: radioControl.RadioControl
});

exports.Radio = radio.Radio;
exports.useRadioContext = radioContext.useRadioContext;
exports.RadioControl = radioControl.RadioControl;
exports.useRadioGroupContext = radioGroupContext.useRadioGroupContext;
exports.RadioGroupLabel = radioGroupLabel.RadioGroupLabel;
exports.RadioLabel = radioLabel.RadioLabel;
Object.defineProperty(exports, 'radioGroupAnatomy', {
  enumerable: true,
  get: () => radio$1.anatomy
});
exports.RadioGroup = RadioGroup;
