// @ark-ui/react/accordion/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const accordion$1 = require('./accordion.cjs');
const accordionContent = require('./accordion-content.cjs');
const accordionContext = require('./accordion-context.cjs');
const accordionItem = require('./accordion-item.cjs');
const accordionItemContext = require('./accordion-item-context.cjs');
const accordionTrigger = require('./accordion-trigger.cjs');
const accordion = require('@zag-js/accordion');

const Accordion = Object.assign(accordion$1.Accordion, {
  Root: accordion$1.Accordion,
  Content: accordionContent.AccordionContent,
  Item: accordionItem.AccordionItem,
  Trigger: accordionTrigger.AccordionTrigger
});

exports.AccordionContent = accordionContent.AccordionContent;
exports.useAccordionContext = accordionContext.useAccordionContext;
exports.AccordionItem = accordionItem.AccordionItem;
exports.useAccordionItemContext = accordionItemContext.useAccordionItemContext;
exports.AccordionTrigger = accordionTrigger.AccordionTrigger;
Object.defineProperty(exports, 'accordionAnatomy', {
  enumerable: true,
  get: () => accordion.anatomy
});
exports.Accordion = Accordion;
