// @ark-ui/react/segment-group/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const segment = require('./segment.cjs');
const segmentContext = require('./segment-context.cjs');
const segmentControl = require('./segment-control.cjs');
const segmentGroup = require('./segment-group.cjs');
const segmentGroupContext = require('./segment-group-context.cjs');
const segmentGroupIndicator = require('./segment-group-indicator.cjs');
const segmentGroupLabel = require('./segment-group-label.cjs');
const segmentGroup_anatomy = require('./segment-group.anatomy.cjs');
const segmentLabel = require('./segment-label.cjs');

const SegmentGroup = Object.assign(segmentGroup.SegmentGroup, {
  Root: segmentGroup.SegmentGroup,
  Label: segmentGroupLabel.SegmentGroupLabel,
  Indicator: segmentGroupIndicator.SegmentGroupIndicator,
  Segment: segment.Segment,
  SegmentLabel: segmentLabel.SegmentLabel,
  SegmentControl: segmentControl.SegmentControl
});

exports.Segment = segment.Segment;
exports.useSegmentContext = segmentContext.useSegmentContext;
exports.SegmentControl = segmentControl.SegmentControl;
exports.useSegmentGroupContext = segmentGroupContext.useSegmentGroupContext;
exports.SegmentGroupIndicator = segmentGroupIndicator.SegmentGroupIndicator;
exports.SegmentGroupLabel = segmentGroupLabel.SegmentGroupLabel;
exports.segmentGroupAnatomy = segmentGroup_anatomy.segmentGroupAnatomy;
exports.SegmentLabel = segmentLabel.SegmentLabel;
exports.SegmentGroup = SegmentGroup;
