// @ark-ui/react/slider/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const slider$1 = require('./slider.cjs');
const sliderContext = require('./slider-context.cjs');
const sliderControl = require('./slider-control.cjs');
const sliderLabel = require('./slider-label.cjs');
const sliderMarker = require('./slider-marker.cjs');
const sliderMarkerGroup = require('./slider-marker-group.cjs');
const sliderOutput = require('./slider-output.cjs');
const sliderRange = require('./slider-range.cjs');
const sliderThumb = require('./slider-thumb.cjs');
const sliderTrack = require('./slider-track.cjs');
const slider = require('@zag-js/slider');

const Slider = Object.assign(slider$1.Slider, {
  Root: slider$1.Slider,
  Control: sliderControl.SliderControl,
  Label: sliderLabel.SliderLabel,
  Marker: sliderMarker.SliderMarker,
  MarkerGroup: sliderMarkerGroup.SliderMarkerGroup,
  Output: sliderOutput.SliderOutput,
  Range: sliderRange.SliderRange,
  Thumb: sliderThumb.SliderThumb,
  Track: sliderTrack.SliderTrack
});

exports.useSliderContext = sliderContext.useSliderContext;
exports.SliderControl = sliderControl.SliderControl;
exports.SliderLabel = sliderLabel.SliderLabel;
exports.SliderMarker = sliderMarker.SliderMarker;
exports.SliderMarkerGroup = sliderMarkerGroup.SliderMarkerGroup;
exports.SliderOutput = sliderOutput.SliderOutput;
exports.SliderRange = sliderRange.SliderRange;
exports.SliderThumb = sliderThumb.SliderThumb;
exports.SliderTrack = sliderTrack.SliderTrack;
Object.defineProperty(exports, 'sliderAnatomy', {
  enumerable: true,
  get: () => slider.anatomy
});
exports.Slider = Slider;
