// @inkeep/preset-chakra/dist/vars.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/vars.ts
var vars_exports = {};
__export(vars_exports, {
  cssVariables: () => cssVariables
});
module.exports = __toCommonJS(vars_exports);
var import_shared2 = require("@inkeep/shared");

// src/recipes/tooltip.recipe.ts
var import_anatomy = require("@ark-ui/anatomy");
var import_shared = require("@inkeep/shared");
var import_dev = require("@pandacss/dev");
var vars = import_shared.cssVar.scope("tooltip", ["bg", "fg", "arrow-bg"]);
var arrow = import_shared.cssVar.scope("arrow", [
  "size",
  "size-half",
  "background",
  "offset"
]);
var tooltipRecipe = (0, import_dev.defineSlotRecipe)({
  className: "tooltip",
  slots: import_anatomy.tooltipAnatomy.keys(),
  base: {
    positioner: {
      [vars["arrow-bg"].name]: vars.bg.ref,
      [vars.bg.name]: "colors.gray.700",
      [vars.fg.name]: "colors.whiteAlpha.900",
      _dark: {
        [vars.bg.name]: "colors.gray.300",
        [vars.fg.name]: "colors.gray.900"
      }
    },
    content: {
      bg: vars.bg.ref,
      color: vars.fg.ref,
      px: "2",
      py: "0.5",
      borderRadius: "sm",
      fontWeight: "medium",
      fontSize: "sm",
      boxShadow: "md",
      maxW: "xs",
      zIndex: "tooltip",
      _open: {
        animation: "fade-in 0.25s ease-out"
      },
      _closed: {
        animation: "fade-out 0.2s ease-out"
      }
    },
    arrow: {
      // this is from Ark-UI/ZagJS
      // https://github.com/chakra-ui/zag/blob/9a3a82f0b3738beda59c313fafd51360e6b0322f/website/data/components/tooltip.mdx#L261
      [arrow.size.name]: "sizes.3",
      [arrow.background.name]: vars["arrow-bg"].ref
    },
    arrowTip: {
      _open: {
        animation: "fade-in 0.25s ease-out"
      },
      _closed: {
        animation: "fade-out 0.2s ease-out"
      }
    }
  }
});

// src/vars.ts
var cssVariables = {
  skeleton: import_shared2.cssVar.scope("skeleton", ["start-color", "end-color"]),
  tooltip: vars
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  cssVariables
});
