// @ark-ui/react/pagination/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const pagination = require('./pagination.cjs');
const paginationContext = require('./pagination-context.cjs');
const paginationEllipsis = require('./pagination-ellipsis.cjs');
const paginationList = require('./pagination-list.cjs');
const paginationListItem = require('./pagination-list-item.cjs');
const paginationNextPageTrigger = require('./pagination-next-page-trigger.cjs');
const paginationPageTrigger = require('./pagination-page-trigger.cjs');
const paginationPrevPageTrigger = require('./pagination-prev-page-trigger.cjs');
const pagination_anatomy = require('./pagination.anatomy.cjs');

const Pagination = Object.assign(pagination.Pagination, {
  Root: pagination.Pagination,
  Ellipsis: paginationEllipsis.PaginationEllipsis,
  List: paginationList.PaginationList,
  ListItem: paginationListItem.PaginationListItem,
  NextPageTrigger: paginationNextPageTrigger.PaginationNextPageTrigger,
  PageTrigger: paginationPageTrigger.PaginationPageTrigger,
  PrevPageTrigger: paginationPrevPageTrigger.PaginationPrevPageTrigger
});

exports.usePaginationContext = paginationContext.usePaginationContext;
exports.PaginationEllipsis = paginationEllipsis.PaginationEllipsis;
exports.PaginationList = paginationList.PaginationList;
exports.PaginationListItem = paginationListItem.PaginationListItem;
exports.PaginationNextPageTrigger = paginationNextPageTrigger.PaginationNextPageTrigger;
exports.PaginationPageTrigger = paginationPageTrigger.PaginationPageTrigger;
exports.PaginationPrevPageTrigger = paginationPrevPageTrigger.PaginationPrevPageTrigger;
exports.paginationAnatomy = pagination_anatomy.paginationAnatomy;
exports.Pagination = Pagination;
