// @ark-ui/react/pagination/index.mjs
'use client';
import { Pagination as Pagination$1 } from './pagination.mjs';
export { usePaginationContext } from './pagination-context.mjs';
import { PaginationEllipsis } from './pagination-ellipsis.mjs';
import { PaginationList } from './pagination-list.mjs';
import { PaginationListItem } from './pagination-list-item.mjs';
import { PaginationNextPageTrigger } from './pagination-next-page-trigger.mjs';
import { PaginationPageTrigger } from './pagination-page-trigger.mjs';
import { PaginationPrevPageTrigger } from './pagination-prev-page-trigger.mjs';
export { paginationAnatomy } from './pagination.anatomy.mjs';

const Pagination = Object.assign(Pagination$1, {
  Root: Pagination$1,
  Ellipsis: PaginationEllipsis,
  List: PaginationList,
  ListItem: PaginationListItem,
  NextPageTrigger: PaginationNextPageTrigger,
  PageTrigger: PaginationPageTrigger,
  PrevPageTrigger: PaginationPrevPageTrigger
});

export { Pagination, PaginationEllipsis, PaginationList, PaginationListItem, PaginationNextPageTrigger, PaginationPageTrigger, PaginationPrevPageTrigger };
