// lil-fp/dist/arr.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from2, except, desc) => {
  if (from2 && typeof from2 == "object" || typeof from2 == "function")
    for (let key of __getOwnPropNames(from2))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from2[key], enumerable: !(desc = __getOwnPropDesc(from2, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);

// src/arr.ts
var arr_exports = {};
__export(arr_exports, {
  at: () => at,
  compact: () => compact,
  concat: () => concat,
  every: () => every,
  filter: () => filter,
  filterMap: () => filterMap,
  flat: () => flat,
  forEach: () => forEach,
  from: () => from,
  has: () => has,
  head: () => head,
  map: () => map,
  next: () => next,
  prev: () => prev,
  reduce: () => reduce,
  some: () => some,
  tail: () => tail,
  uniq: () => uniq
});
module.exports = __toCommonJS(arr_exports);
var from = (v) => Array.from(v), filter = (f) => (arr) => arr.filter(f), map = (f) => (arr) => arr.map(f), reduce = (f, init) => (arr) => arr.reduce(f, init), filterMap = (f) => (arr) => arr.reduce((acc, v, index, arr2) => {
  let x = f(v, index, arr2);
  return x == null ? acc : [...acc, x];
}, []), forEach = (f) => (arr) => arr.forEach(f), next = (i, loop) => (arr) => arr[(i + 1) % (loop ? arr.length : arr.length - 1)], prev = (i, loop) => (arr) => arr[(i - 1 + (loop ? arr.length : 0)) % arr.length], head = (arr) => arr[0], tail = (arr) => arr[arr.length - 1], at = (i) => (arr) => arr.at(i), uniq = (arr) => [...new Set(arr)], some = (f) => (arr) => arr.some(f), every = (f) => (arr) => arr.every(f), has = (v) => (arr) => arr.includes(v), concat = (...others) => (v) => v.concat(...others), compact = (arr) => arr.filter((v) => v != null), flat = () => (arr) => arr.flat();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  at,
  compact,
  concat,
  every,
  filter,
  filterMap,
  flat,
  forEach,
  from,
  has,
  head,
  map,
  next,
  prev,
  reduce,
  some,
  tail,
  uniq
});
