// lil-fp/dist/chunk-5IWUBL4K.mjs
import {
  __export
} from "./chunk-LLLY6SAS.mjs";

// src/opt.ts
var opt_exports = {};
__export(opt_exports, {
  filter: () => filter,
  flatMap: () => flatMap,
  fromExecution: () => fromExecution,
  fromNullable: () => fromNullable,
  fromPredicate: () => fromPredicate,
  getOrElse: () => getOrElse,
  getOrThrow: () => getOrThrow,
  isNone: () => isNone,
  isSome: () => isSome,
  map: () => map,
  match: () => match,
  none: () => none,
  orElse: () => orElse,
  some: () => some,
  tap: () => tap
});
var none = Object.freeze({ _tag: "None" }), some = (value) => Object.freeze({ _tag: "Some", value });
function fromPredicate(predicate) {
  return (v) => predicate(v) ? some(v) : none;
}
var fromNullable = (value) => value == null ? none : some(value), fromExecution = (value) => {
  try {
    return fromNullable(value());
  } catch {
    return none;
  }
}, isNone = (o) => o._tag === "None", isSome = (o) => o._tag === "Some", isOption = (o) => o._tag === "None" || o._tag === "Some", map = (f) => (o) => isNone(o) ? none : some(f(o.value)), flatMap = (f) => (o) => isNone(o) ? none : f(o.value), getOrElse = (def) => (o) => isNone(o) ? def() : o.value, getOrThrow = (msg) => (o) => {
  if (isNone(o))
    throw new Error(msg);
  return o.value;
}, match = (onSome, onNone) => (o) => isNone(o) ? onNone() : onSome(o.value), orElse = (v) => (o) => {
  let vv = v();
  return isOption(vv) ? isNone(o) ? vv : o : isNone(o) ? some(vv) : o;
}, tap = (f) => (o) => (isSome(o) && f(o.value), o), filter = (f) => (o) => isNone(o) ? none : f(o.value) ? o : none;

export {
  none,
  some,
  fromPredicate,
  fromNullable,
  fromExecution,
  isNone,
  isSome,
  map,
  flatMap,
  getOrElse,
  getOrThrow,
  match,
  orElse,
  tap,
  filter,
  opt_exports
};
