// lil-fp/dist/is.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: !0 });
}, __copyProps = (to, from, except, desc) => {
  if (from && typeof from == "object" || typeof from == "function")
    for (let key of __getOwnPropNames(from))
      !__hasOwnProp.call(to, key) && key !== except && __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: !0 }), mod);

// src/is.ts
var is_exports = {};
__export(is_exports, {
  isArr: () => isArr,
  isBool: () => isBool,
  isFunc: () => isFunc,
  isNum: () => isNum,
  isObj: () => isObj,
  isStr: () => isStr
});
module.exports = __toCommonJS(is_exports);
var isBool = (x) => typeof x == "boolean", isNum = (x) => typeof x == "number", isStr = (x) => typeof x == "string", isArr = (x) => Array.isArray(x), isObj = (x) => x !== null && typeof x == "object" && !isArr(x), isFunc = (x) => typeof x == "function";
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  isArr,
  isBool,
  isFunc,
  isNum,
  isObj,
  isStr
});
