// @ark-ui/react/rating-group/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const rating = require('./rating.cjs');
const ratingContext = require('./rating-context.cjs');
const ratingGroup$1 = require('./rating-group.cjs');
const ratingGroupContext = require('./rating-group-context.cjs');
const ratingGroupControl = require('./rating-group-control.cjs');
const ratingGroupLabel = require('./rating-group-label.cjs');
const ratingGroup = require('@zag-js/rating-group');

const RatingGroup = Object.assign(ratingGroup$1.RatingGroup, {
  Root: ratingGroup$1.RatingGroup,
  Control: ratingGroupControl.RatingGroupControl,
  Label: ratingGroupLabel.RatingGroupLabel,
  Rating: rating.Rating
});

exports.Rating = rating.Rating;
exports.useRatingContext = ratingContext.useRatingContext;
exports.useRatingGroupContext = ratingGroupContext.useRatingGroupContext;
exports.RatingGroupControl = ratingGroupControl.RatingGroupControl;
exports.RatingGroupLabel = ratingGroupLabel.RatingGroupLabel;
Object.defineProperty(exports, 'ratingGroupAnatomy', {
  enumerable: true,
  get: () => ratingGroup.anatomy
});
exports.RatingGroup = RatingGroup;
