// @ark-ui/react/toast/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const toast = require('./toast.cjs');
const toastCloseTrigger = require('./toast-close-trigger.cjs');
const toastDescription = require('./toast-description.cjs');
const toastGroup = require('./toast-group.cjs');
const toastItemContext = require('./toast-item-context.cjs');
const toastPlacements = require('./toast-placements.cjs');
const toastProvider = require('./toast-provider.cjs');
const toastTitle = require('./toast-title.cjs');
const toast$1 = require('@zag-js/toast');

const Toast = Object.assign(toast.Toast, {
  Provider: toastProvider.ToastProvider,
  Title: toastTitle.ToastTitle,
  Root: toast.Toast,
  Description: toastDescription.ToastDescription,
  CloseTrigger: toastCloseTrigger.ToastCloseTrigger,
  Placements: toastPlacements.ToastPlacements,
  Group: toastGroup.ToastGroup
});

exports.ToastRoot = toast.Toast;
exports.ToastCloseTrigger = toastCloseTrigger.ToastCloseTrigger;
exports.ToastDescription = toastDescription.ToastDescription;
exports.ToastGroup = toastGroup.ToastGroup;
exports.useToastItemContext = toastItemContext.useToastItemContext;
exports.ToastPlacements = toastPlacements.ToastPlacements;
exports.ToastProvider = toastProvider.ToastProvider;
exports.useToast = toastProvider.useToast;
exports.ToastTitle = toastTitle.ToastTitle;
Object.defineProperty(exports, 'toastAnatomy', {
  enumerable: true,
  get: () => toast$1.anatomy
});
exports.Toast = Toast;
