// @ark-ui/react/dialog/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const dialog$1 = require('./dialog.cjs');
const dialogBackdrop = require('./dialog-backdrop.cjs');
const dialogCloseTrigger = require('./dialog-close-trigger.cjs');
const dialogContainer = require('./dialog-container.cjs');
const dialogContent = require('./dialog-content.cjs');
const dialogContext = require('./dialog-context.cjs');
const dialogDescription = require('./dialog-description.cjs');
const dialogTitle = require('./dialog-title.cjs');
const dialogTrigger = require('./dialog-trigger.cjs');
const dialog = require('@zag-js/dialog');

const Dialog = Object.assign(dialog$1.Dialog, {
  Root: dialog$1.Dialog,
  Backdrop: dialogBackdrop.DialogBackdrop,
  CloseTrigger: dialogCloseTrigger.DialogCloseTrigger,
  Container: dialogContainer.DialogContainer,
  Content: dialogContent.DialogContent,
  Description: dialogDescription.DialogDescription,
  Title: dialogTitle.DialogTitle,
  Trigger: dialogTrigger.DialogTrigger
});

exports.DialogBackdrop = dialogBackdrop.DialogBackdrop;
exports.DialogCloseTrigger = dialogCloseTrigger.DialogCloseTrigger;
exports.DialogContainer = dialogContainer.DialogContainer;
exports.DialogContent = dialogContent.DialogContent;
exports.useDialogContext = dialogContext.useDialogContext;
exports.DialogDescription = dialogDescription.DialogDescription;
exports.DialogTitle = dialogTitle.DialogTitle;
exports.DialogTrigger = dialogTrigger.DialogTrigger;
Object.defineProperty(exports, 'dialogAnatomy', {
  enumerable: true,
  get: () => dialog.anatomy
});
exports.Dialog = Dialog;
