// @pandacss/postcss/dist/index.mjs
// ../../node_modules/.pnpm/tsup@7.1.0_typescript@5.3.3/node_modules/tsup/assets/esm_shims.js
import { fileURLToPath } from "url";
import path from "path";
var getFilename = () => fileURLToPath(import.meta.url);
var getDirname = () => path.dirname(getFilename());
var __dirname = /* @__PURE__ */ getDirname();

// src/index.ts
import { Builder } from "@pandacss/node";
import { createRequire } from "module";
import path2 from "path";
var customRequire = createRequire(__dirname);
var PLUGIN_NAME = "pandacss";
var interopDefault = (obj) => obj && obj.__esModule ? obj.default : obj;
var loadConfig = () => interopDefault(customRequire("@pandacss/postcss"));
var builder = new Builder();
var nodeModulesRegex = /node_modules/;
function isValidCss(file) {
  const [filePath] = file.split("?");
  return path2.extname(filePath) === ".css";
}
var shouldSkip = (fileName) => {
  if (!fileName)
    return true;
  if (!isValidCss(fileName))
    return true;
  if (fileName.includes("@pandacss/astro/base.css"))
    return false;
  return nodeModulesRegex.test(fileName);
};
var pandacss = (options = {}) => {
  const { configPath, cwd } = options;
  return {
    postcssPlugin: PLUGIN_NAME,
    plugins: [
      async function(root, result) {
        const fileName = result.opts.from;
        const skip = shouldSkip(fileName);
        if (skip)
          return;
        await builder.setup({ configPath, cwd });
        if (!builder.isValidRoot(root))
          return;
        await builder.emit();
        await builder.extract();
        builder.registerDependency((dep) => {
          result.messages.push({
            ...dep,
            plugin: PLUGIN_NAME,
            parent: result.opts.from
          });
        });
        builder.write(root);
        root.walk((node) => {
          if (!node.source) {
            node.source = root.source;
          }
        });
      }
    ]
  };
};
pandacss.postcss = true;
var src_default = pandacss;
export {
  src_default as default,
  loadConfig,
  pandacss
};
