// @ark-ui/react/pin-input/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const pinInput$1 = require('./pin-input.cjs');
const pinInputContext = require('./pin-input-context.cjs');
const pinInputControl = require('./pin-input-control.cjs');
const pinInputField = require('./pin-input-field.cjs');
const pinInputLabel = require('./pin-input-label.cjs');
const pinInput = require('@zag-js/pin-input');

const PinInput = Object.assign(pinInput$1.PinInput, {
  Root: pinInput$1.PinInput,
  Control: pinInputControl.PinInputControl,
  Input: pinInputField.PinInputInput,
  Label: pinInputLabel.PinInputLabel
});

exports.usePinInputContext = pinInputContext.usePinInputContext;
exports.PinInputControl = pinInputControl.PinInputControl;
exports.PinInputInput = pinInputField.PinInputInput;
exports.PinInputLabel = pinInputLabel.PinInputLabel;
Object.defineProperty(exports, 'pinInputAnatomy', {
  enumerable: true,
  get: () => pinInput.anatomy
});
exports.PinInput = PinInput;
