// @ark-ui/react/carousel/index.cjs
'use client';
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const carousel = require('./carousel.cjs');
const carouselContext = require('./carousel-context.cjs');
const carouselControl = require('./carousel-control.cjs');
const carouselIndicator = require('./carousel-indicator.cjs');
const carouselIndicatorGroup = require('./carousel-indicator-group.cjs');
const carouselNextSlideTrigger = require('./carousel-next-slide-trigger.cjs');
const carouselPrevSlideTrigger = require('./carousel-prev-slide-trigger.cjs');
const carouselSlide = require('./carousel-slide.cjs');
const carouselSlideGroup = require('./carousel-slide-group.cjs');
const carouselViewport = require('./carousel-viewport.cjs');
const carousel_anatomy = require('./carousel.anatomy.cjs');

const Carousel = Object.assign(carousel.Carousel, {
  Root: carousel.Carousel,
  Control: carouselControl.CarouselControl,
  Indicator: carouselIndicator.CarouselIndicator,
  IndicatorGroup: carouselIndicatorGroup.CarouselIndicatorGroup,
  NextSlideTrigger: carouselNextSlideTrigger.CarouselNextSlideTrigger,
  PrevSlideTrigger: carouselPrevSlideTrigger.CarouselPrevSlideTrigger,
  Slide: carouselSlide.CarouselSlide,
  SlideGroup: carouselSlideGroup.CarouselSlideGroup,
  Viewport: carouselViewport.CarouselViewport
});

exports.useCarouselContext = carouselContext.useCarouselContext;
exports.CarouselControl = carouselControl.CarouselControl;
exports.CarouselIndicator = carouselIndicator.CarouselIndicator;
exports.CarouselIndicatorGroup = carouselIndicatorGroup.CarouselIndicatorGroup;
exports.CarouselNextSlideTrigger = carouselNextSlideTrigger.CarouselNextSlideTrigger;
exports.CarouselPrevSlideTrigger = carouselPrevSlideTrigger.CarouselPrevSlideTrigger;
exports.CarouselSlide = carouselSlide.CarouselSlide;
exports.CarouselSlideGroup = carouselSlideGroup.CarouselSlideGroup;
exports.CarouselViewport = carouselViewport.CarouselViewport;
exports.carouselAnatomy = carousel_anatomy.carouselAnatomy;
exports.Carousel = Carousel;
